Add orientation transforms to stream

uvjustin opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
As stream is fmp4/mp4 based, we can apply affine transformations without transcoding by simply modifying the transformation matrix in the mp4 metadata.
The still images generated by stream have always involved decoding a keyframe and reencoding it to jpeg. The additional work of applying a transformation is minimal as it only involves adding a simple numpy operation to the pipeline.
This PR updates stream to change the orientation of video and the generated still images using the 8 EXIF orientations as guidelines. Note that for video, the application of these transforms is player dependent and may only work in browser for now. Also, orientations 5 and 7 do not appear to work properly for video.
This PR also adds an `orientation` field to `StreamSettings` and exposes this as a property in `Stream`.
In `Camera`, an `orientation` property is added to `CameraEntityPreferences` and this is exposed to the frontend via the `camera/update_prefs` websocket endpoint. This is essentially how `preload_stream` is currently stored/handled.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

uvjustin committed:
Add orientation transforms to stream

uvjustin commented:
It might be preferable to expose/store this option similar to how keepalive is exposed. That way the values can be updated more easily and on the fly, and we also don't have to plumb it through an option on every integration that uses stream.

uvjustin committed:
Use camera prefs to set orientation

uvjustin commented:
Added [a frontend PR](https://github.com/home-assistant/frontend/pull/13512) to use this feature.

uvjustin committed:
Add missing setter comment

uvjustin committed:
Modify mp4 transform matrix lazily

uvjustin committed:
Add stream recorder test

emontnemery commented:
Not directly related to this PR, but maybe `camera`'s custom store should be deprecated in favor of entity registry options which is used for similar purposes by `number`, `sensor` and `weather`.

uvjustin commented:
> Not directly related to this PR, but maybe `camera`'s custom store should be deprecated in favor of entity registry options which is used for similar purposes by `number`, `sensor` and `weather`?

I wasn't familiar with the entity registry options but had a look...doesn't that require the camera entities to all have unique IDs?

uvjustin committed:
Use entity registry to store orientation preferences

uvjustin commented:
@emontnemery Since many cameras still don't use config flows, switching to the entity registry doesn't seem like an option.
However, since this is a new feature, we can restrict its use to entities in the entity registry. I've pushed a commit which stores the orientation preferences in the entity registry options, separately from the preload stream preferences. The preload stream preferences can be migrated another time.

uvjustin committed:
Add test coverage

allenporter commented on tests/components/camera/test_init.py:
```diff
@@ -294,29 +294,73 @@ async def test_websocket_get_prefs(hass, hass_ws_client, mock_camera):
+    assert legacy_camera_prefs[PREF_PRELOAD_STREAM] is True
+    assert response["result"][PREF_PRELOAD_STREAM] is True
+
+    # Try sending orientation update for entity not in entity registry
```
I suggest splitting this up into a few different tests (e.g. with unique id, without unique id, etc) -- just for making it easier to follow exactly what state is under test for each scenario.

uvjustin replied:
Ended up splitting it into two tests, one for preload and one for orientation (the case where both values are updated simultaneously is no longer exercised, but the frontend does not send that anyway).

allenporter commented on tests/components/camera/test_init.py:
```diff
@@ -294,29 +294,73 @@ async def test_websocket_get_prefs(hass, hass_ws_client, mock_camera):
-        == setup_camera_prefs[PREF_PRELOAD_STREAM]
+
+    # preload_stream entry for this camera should have been added
+    assert "camera.demo_uniquecamera" in legacy_camera_prefs_by_entity
```
It is typically preferred to not dig into `hass.data` during a test. Can you use the websocket to read prefs to verify pref state?

allenporter reviewed (changes requested):
This is really cool.  A few comments below...

uvjustin committed:
Use exception for camera preference update failure

uvjustin committed:
Use lookup tuple for transform function
Add comments to transform functions
Modify operations in 5 and 7 for consistency

uvjustin committed:
Use tuple instead of list for transform matrix lookup

uvjustin committed:
Avoid using hass.data in update_prefs test

uvjustin committed:
Split update_prefs tests

uvjustin committed:
Revert recorder container_options change

uvjustin committed:
Remove attribution

uvjustin commented on homeassistant/components/stream/core.py:
```diff
@@ -430,6 +434,28 @@ def create_codec_context(self, codec_context: CodecContext) -> None:
+
+        Adapted from https://github.com/lilohuang/PyTurboJPEG.
+        """
+        if orientation == 1:
```
I ended up moving the transform functions into a tuple for fast lookup, but I put the comments in there.

uvjustin commented on homeassistant/components/stream/recorder.py:
```diff
@@ -104,7 +105,11 @@ def write_segment(segment: Segment) -> None:
-                        "video_track_timescale": str(int(1 / source_v.time_base))
+                        "video_track_timescale": str(int(1 / source_v.time_base)),
+                        "movflags": "frag_keyframe",
+                        "min_frag_duration": str(
```
You got me. We can remove this and submit it as a separate PR.

uvjustin committed:
Fix comment in test

MartinHjelmare commented on homeassistant/components/camera/__init__.py:
```diff
@@ -888,9 +891,12 @@ async def websocket_update_prefs(
-    connection.send_result(msg["id"], prefs.get(entity_id).as_dict())
+    try:
+        entity_prefs = await prefs.async_update(entity_id, **changes)
+        connection.send_result(msg["id"], entity_prefs)
```
Please only wrap the line that we expect to raise in the try... except block. Sending the result shouldn't raise, right?

uvjustin replied:
Thanks Martin. I'll make a follow up PR to change this and the comment below.

MartinHjelmare commented on homeassistant/components/camera/prefs.py:
```diff
@@ -53,22 +63,36 @@ async def async_update(
+    ) -> dict[str, bool | int]:
+        """Update camera preferences.
+
+        Returns a dict with the preferences on success or a string on error.
```
It doesn't return a string on error.

bramkragten commented:
We should move these settings (both preload and orientation) to the entity registry, as we now have `options` in the entity registry

uvjustin commented:
I don't quite follow. This PR already puts the new orientation options in the entity registry.
As mentioned above, it doesn't seem like preload can be moved into the entity registry as many camera integrations still don't use unique IDs.

bramkragten commented:
OK, sorry then. Then I would suggest to use the normal api's to update the entity registry instead of `camera/update_prefs`
